### DISCUSSION TITLE
Generalised Ref forms

### DISCUSSION THREAD

**Discussion by kiranandcode** [racket/rhombus#307] [Ideas]
It would be nice if the following worked:

```
#lang rhombus

class Example():
  implements Refable
  override method ref(index, ...):
    println([index, ...])

Example()[10,20]
```

**Comment by mflatt** [comment]
I think this generalization — allowing multiple dimensions for a ref —could be made to work. When static information is available, the information could be used to make there a suitable number of dimensions are used (i.e., the right number of comma-separated subexpressions within `[]`).

In that case that static information is not available, though, things that are currently rejected statically would turn into dynamic errors. For example, `x[1, 2]` is currently a syntax error if `x` is a variable or an expression without any particular static information, but it would be allowed, and then there would be a dynamic error if `x` is a typical indexable value like a list, array, or map.

So, there's a tradeoff between supporting the generalization and providing better and earlier errors for the common 1-dimensional case. If multi-dimensionsal cases will be rare, then it might be better to leave that to another path, such as a `ref` method that accepts more arguments or using a list of indices as an index.

I am not immediately sure of the right choice here.

(All of this is about the default `#%ref`. A variant of Rhombus can always use a different `#%ref`, either more limited or more general.)

**Comment by jackfirth** [comment]
I'm not all that into this particular generalization because it doesn't buy much over just writing `a[x][y]`. It's only one character more.

**Comment by mflatt** [comment]
Another idea: Rhombus could provide a variant of `#%ref` (in dynamic and static forms) that allows multiple dimensions. So, in contexts where multiple dimensions are desirable, the extended version of `#%ref` could be imported. But by default, behavior and reporting for `[]` refs stays as it is now.
